openvpn/middlewares/server/credentials/middleware.go

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2020 The "MysteriumNetwork/go-openvpn" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package credentials
+
+import (
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/log"
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server"
+	"github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth"
+)
+
+// Middleware is able to process authorize incoming clients by given credentials validator callback.
+type Middleware struct {
+	*auth.Middleware
+
+	validator Validator
+}
+
+// Validator callback checks given auth primitives.
+type Validator func(clientID int, username, password string) (bool, error)
+
+// NewMiddleware creates server user_auth challenge authentication Middleware
+func NewMiddleware(validator Validator) *Middleware {
+	m := &Middleware{
+		validator: validator,
+	}
+	m.Middleware = auth.NewMiddleware(m.handleClientEvent)
```

Comment:
I failed to parse these recursive dependencies. Can we make it more clear?
```suggestion
	m.Middleware = auth.NewMiddleware(m.handleClientEvent)
```

Reply:
It's simple inheritance. Improved constructor